server/mailrepository/mailrepository-cassandra/src/main/java/org/apache/james/mailrepository/cassandra/CassandraMailRepository.java

```diff
@@ -19,48 +19,195 @@
+import java.nio.ByteBuffer;
 import java.util.Collection;
 import java.util.Iterator;
+import java.util.Properties;
+import java.util.concurrent.CompletableFuture;
 
 import javax.mail.MessagingException;
+import javax.mail.Session;
+import javax.mail.internet.MimeMessage;
 
+import org.apache.commons.io.IOUtils;
+import org.apache.commons.lang3.tuple.Pair;
+import org.apache.james.blob.api.BlobId;
+import org.apache.james.blob.api.ObjectStore;
 import org.apache.james.mailrepository.api.MailRepository;
+import org.apache.james.util.BodyOffsetInputStream;
+import org.apache.james.util.CompletableFutureUtil;
+import org.apache.james.util.FluentFutureStream;
 import org.apache.mailet.Mail;
 
+import com.github.fge.lambdas.Throwing;
+import com.google.common.base.Throwables;
+import com.google.common.primitives.Bytes;
+
 public class CassandraMailRepository implements MailRepository {
 
-    public CassandraMailRepository() {
+    private final String url;
+    private final CassandraMailRepositoryKeysDAO keysDAO;
+    private final CassandraMailRepositoryCountDAO countDAO;
+    private final CassandraMailRepositoryMailDAO mailDAO;
+    private final ObjectStore objectStore;
+
+    public CassandraMailRepository(String url, CassandraMailRepositoryKeysDAO keysDAO, CassandraMailRepositoryCountDAO countDAO, CassandraMailRepositoryMailDAO mailDAO, ObjectStore objectStore) {
+        this.url = url;
+        this.keysDAO = keysDAO;
+        this.countDAO = countDAO;
+        this.mailDAO = mailDAO;
+        this.objectStore = objectStore;
     }
 
     @Override
-    public void store(Mail mail) {
+    public void store(Mail mail) throws MessagingException {
+        try {
+            Pair<byte[], byte[]> splitHeaderBody = splitHeaderBody(mail.getMessage());
+
+            CompletableFuture<Pair<BlobId, BlobId>> blobIds = CompletableFutureUtil.combine(
+                objectStore.save(splitHeaderBody.getLeft()),
+                objectStore.save(splitHeaderBody.getRight()),
+                Pair::of);
+
+            blobIds.thenCompose(Throwing.function(pair ->
+                mailDAO.store(url, mail, pair.getLeft(), pair.getRight())))
+                .thenCompose(any -> CompletableFuture.allOf(
+                    countDAO.increment(url),
+                    keysDAO.store(url, mail.getName())))
+                .join();
+        } catch (IOException e) {
+            throw new MessagingException("Exception while storing mail", e);
+        }
+    }
+
+    public Pair<byte[], byte[]> splitHeaderBody(MimeMessage message) throws IOException, MessagingException {
+        byte[] messageAsArray = messageToArray(message);
+        int bodyStartOctet = computeBodyStartOctet(messageAsArray);
+
+        return Pair.of(
+            getHeaderBytes(messageAsArray, bodyStartOctet),
+            getBodyBytes(messageAsArray, bodyStartOctet));
+    }
+
+    public byte[] messageToArray(MimeMessage message) throws IOException, MessagingException {
```

Comment:
all these methods probably do not belong to CassandraMailRepository, they are of general purpose.
Also, the raw array handling is from another decade, we shoud be able to find modern inputstream primitives instead.